R/prepare_isf_2023_scenario.R

```diff
@@ -0,0 +1,306 @@
+#' Prepare ISF 2023 scenario data
+#'
+#' @param isf_2023_scope_global_raw A tidyxl data frame (with a `formats`
+#'   attribute) with a raw ISF Scope Global 2023 import.
+#' @param isf_2023_s_global_raw A tidyxl data frame (with a `formats` attribute)
+#'   with a raw ISF S_Global 2023 import.
+#' @param isf_2023_annex_countries_raw A list of tidyxl data frames (with a
+#'   `formats` attribute) containing the raw import of each of the Annex
+#'   Countries xlsx files for ISF 2023.
+#'
+#' @return A prepared ISF 2023 scenario data-frame.
+#'
+#' @importFrom dplyr %>%
+#'
+#' @export
+
+prepare_isf_2023_scenario <- function(isf_2023_scope_global_raw,
+                                      isf_2023_s_global_raw,
+                                      isf_2023_annex_countries_raw) {
+  isf_2023_power <-
+    isf_2023_annex_countries_raw %>%
+    purrr::map(extract_installed_capacity) %>%
+    purrr::list_rbind()
+
+  isf_2023_fossil_fuels <-
+    isf_2023_annex_countries_raw %>%
+    purrr::map(extract_final_energy_demand) %>%
+    purrr::list_rbind()
+
+  isf_2023_steel_cement <-
+    extract_steel_cement(
+      isf_2023_scope_global_raw,
+      isf_2023_s_global_raw
+    )
+
+  out <-
+    dplyr::bind_rows(
+      isf_2023_power,
+      isf_2023_fossil_fuels,
+      isf_2023_steel_cement
+    )
+
+  out %>%
+    bridge_technologies(isf_2023_technology_bridge) %>%
+    dplyr::summarize(
+      value = sum(.data[["value"]], na.rm = TRUE),
+      .by = c(
+        "source",
+        "scenario",
+        "scenario_geography",
+        "sector",
+        "technology",
+        "indicator",
+        "units",
+        "year"
+      )
+    ) %>%
+    bridge_geographies(isf_2023_geography_bridge) %>%
+    dplyr::arrange(
+      .data[["scenario_geography"]],
+      .data[["sector"]],
+      .data[["technology"]],
+      .data[["year"]]
+    ) %>%
+    dplyr::relocate(
+      "source",
+      "scenario",
+      "scenario_geography",
+      "sector",
+      "technology",
+      "indicator",
+      "units",
+      "year",
+      "value"
+    )
+}
+
+
+extract_installed_capacity <- function(x) {
+  formats <- attr(x, "formats")
+  top_border <- !is.na(formats$local$border$top$style)
+  left_border <- !is.na(formats$local$border$left$style)
+
+  corners <-
+    x %>%
+    dplyr::filter(top_border[.data[["local_format_id"]]] &
+                    left_border[.data[["local_format_id"]]]) %>%
+    dplyr::filter(col == 17)
+
+  partitions <- unpivotr::partition(x, corners)
+  tables <- dplyr::pull(partitions, "cells")
+
+  tables[[1]] %>%
+    dplyr::filter(!.data[["is_blank"]]) %>%
+    unpivotr::behead("up-left", "table") %>%
+    unpivotr::behead("up", "year", formatters = list(numeric = as.double)) %>%
+    unpivotr::behead("left", "technology") %>%
```

Comment:
NB: lol yeah this whole `tidyxl` + `unpivotr` thing feels a bit over-engineered. But still cool that it exists!

Reply:
power vs. less-complexity ⚖️ 
🤣 

https://nacnudus.github.io/spreadsheet-munging-strategies/
> Tidyxl and unpivotr are much more complicated than readxl, and that’s the point. Tidyxl and unpivotr give you more power and complexity when you need it.

https://nacnudus.github.io/tidyxl/index.html#similar-projects
> [tidyxl](https://github.com/nacnudus/tidyxl) was originally derived from [readxl](https://github.com/tidyverse/readxl) and still contains some of the same code, hence it inherits the GPL-3 licence. [readxl](https://github.com/tidyverse/readxl) is intended for importing tabular data with a single row of column headers, whereas [tidyxl](https://github.com/nacnudus/tidyxl) is more general, and less magic.